File: 380. Insert Delete GetRandom O(1)/Insert Delete GetRandom O(1).java
import java.util.Random;

public class RandomizedSet {
    
    private Map<Integer, Integer> keyMap;
    private Map<Integer, Integer> valueMap;
	Random random;
    int count;

    public RandomizedSet() {
        keyMap = new HashMap<>();
        valueMap = new HashMap<>();
		random = new Random();
    }
    
    public boolean insert(int val) {
        if (keyMap.containsKey(val)) {
            return false;
        } else {
            keyMap.put(val, count);
            valueMap.put(count, val);
            count = keyMap.size();
            return true;
        }
    }
    
    public boolean remove(int val) {
        if (!keyMap.containsKey(val)) {
            return false;
        } else {
            int valueKey = keyMap.get(val);
            keyMap.remove(val);
            if (valueKey != valueMap.size() - 1) {
                valueMap.put(valueKey, valueMap.get(valueMap.size() - 1));
                keyMap.put(valueMap.get(valueMap.size() - 1), valueKey);
                valueMap.remove(valueMap.size() - 1);
            } else {
                valueMap.remove(valueKey);
            }
            count = keyMap.size();
            return true;
        }
    }
    
    public int getRandom() {
        int n = random.nextInt(keyMap.size());
        return valueMap.get(n);
    }
}

File: Problems & Solutions/380. Insert Delete GetRandom O(1)/29 November 2022/Insert Delete GetRandom O(1).cs
public class RandomizedSet {
    
    private Dictionary<int,int> cache; 
    private Random rnd;
    private List<int> data;
    public RandomizedSet() {
        cache = new Dictionary<int,int>();
        rnd = new Random();
        data = new List<int>();
    }
    
    public bool Insert(int val) {
        if(cache.ContainsKey(val))
            return false;
        else
        {
            data.Add(val);
            cache.Add(val,(data.Count-1));
            return true;
        }
    }
    
    public bool Remove(int val) {
          if(!cache.ContainsKey(val))
            return false;
        else
        {
            int lastIndex = (data.Count-1);
            int lastElem = data[lastIndex];
            
            int i = cache[val];
            data[i] = lastElem;
            cache[lastElem] = i;
            
            data.RemoveAt(lastIndex);
            cache.Remove(val);
            
            return true;
        }
            
    }
    
    public int GetRandom() {
        Console.WriteLine(cache.Count);
        return data[rnd.Next(data.Count)];
    }
}